### PR TITLE
✨ New feature : Sticky player, 50% visiblity autoplay, 1:1 Mobile Aspect ratio, Full width player in mobile

### PR DESCRIPTION
**Pull request by gopanisandip**
We have updated our player with couple of new feature like:
- Stucky player on scroll
- 50% visiblity autoplay, On Autoplay video will start only if player is more than 50% visible and if enabled from platfom
- On mobile player will visible in 1:1 aspect ratio
- Full width player on mobile by removing left right margin

Also aspect ratio bug fix

### CODE REVIEW

**Comment by gopanisandip** [comment]
@aghassemi  Can you please take a look of this mowplayer extension changes?

**Comment by gopanisandip** [comment]
@alanorozco  Could you please look at travis-ci error.  Its asking to change some viewport realted function used but i am not finding any alsternative.

**Comment by gopanisandip** [comment]
> @gopanisandip
> 
> Thanks for your contribution.
> 
> > 1. Stucky player on scroll
> 
> Could this use `amp-video-docking` instead?
> 
> > 50% visiblity autoplay, On Autoplay video will start only if player is more than 50% visible and if enabled from platfom
> 
> Could this use the `autoplay` attribute instead?
> 
> > On mobile player will visible in 1:1 aspect ratio
> > Full width player on mobile by removing left right margin
> 
> This breaks the layout system and causes content layout shift. Please remove these features.

@alanorozco 

For 50% visiblity autoplay via autoplay attribute will work when player is 50% visible?  OR it just start playing  when player loads?

Regarding 1:1 mobile aspect ratio

i think we already have approved in past which will change height of player.  Now is just mentioned here  because this is set from our backend.

Regarding full width player.  Is there any ulternate solution?  because is really important for us.

**Comment by alanorozco** [comment]
@gopanisandip 

> For 50% visiblity autoplay via autoplay attribute will work when player is 50% visible? OR it just start playing when player loads?

Yes, this is the `autoplay` behavior given by the `VideoManager`.

> i think we already have approved in past which will change height of player. Now is just mentioned here because this is set from our backend.

It used `attemptChangeHeight` previously, which would only be allowed when the element is outside the viewport and therefore wouldn't cause layout shift. This new change uses `forceChangeHeight`, which would break it otherwise.

> Regarding full width player. Is there any ulternate solution? because is really important for us.

Could this instead be handled on the author's side by setting the component's box to `layout="fill"` inside a responsively sized container?

Because of AMP's sizing and asynchronous loading, any component that resizes itself on build/layout would cause content shift.

**Comment by gopanisandip** [comment]
@alanorozco  Can't we allow this for forceChangeHeight about aspect ratio and other things?  Because this mowplayer is not like other player for public use and all publisher using player  knows about the features.

About other changes going to change it as suggested.

**Comment by gopanisandip** [comment]
@alanorozco  Any update?

**Comment by alanorozco** [comment]
> Can't we allow this for forceChangeHeight about aspect ratio and other things? Because this mowplayer is not like other player for public use and all publisher using player knows about the features.

Unfortunately not, any `forceChangeHeight` will cause content shift.

**Comment by gopanisandip** [comment]
@alanorozco So  i am going to apply new changes as per our discussion.

**Comment by gopanisandip** [comment]
@alanorozco  Can you please check?  I changed forceChangeHeight to attemptChangeHeight

Let me know next step.

Thanks

**Comment by gopanisandip** [comment]
@alanorozco   Any update on this merge?

File: extensions/amp-mowplayer/0.1/amp-mowplayer.js [hunk 1]
```diff
@@ -255,20 +286,237 @@ class AmpMowplayer extends AMP.BaseElement {
...
+      if (previousVisible !== visible) {
+        this.sendMessage_('visibility_observer_visibility', {
+          visible,
+        });
+        previousVisible = visible;
+      }
+    };
+    window.addEventListener('scroll', worker);
+    window.addEventListener('resize', worker);
+    worker();
```
Comment (alanorozco) [inline extensions/amp-mowplayer/0.1/amp-mowplayer.js:519]: 1. `getBoundingClientRect` causes reflow. Use `element.getLayoutBox()` instead.
2. `window.innerHeight` also causes reflow. Use the `Viewport` service instead.
3. Don't attach `scroll/resize` events directly to `window. Use the `Viewport` service instead.

This also is very slow/causes jank when executed on every `scroll` event. Instead, let the frame ask for it imperatively, [like `amp-video-iframe`](https://github.com/ampproject/amphtml/blob/master/extensions/amp-video-iframe/0.1/amp-video-iframe.js#L361).

File: extensions/amp-mowplayer/0.1/amp-mowplayer.js [hunk 2]
```diff
@@ -255,20 +286,237 @@ class AmpMowplayer extends AMP.BaseElement {
...
+      window.addEventListener('resize', repaint);
     }
+  }
+
+  /**
+   * Initialize sticky feature
+   * @private
+   * @param {Object} data
+   */
+  onStickPlayer_(data) {
```
Comment (alanorozco) [inline extensions/amp-mowplayer/0.1/amp-mowplayer.js:341]: Please remove this feature and use [`amp-video-docking`](https://go.amp.dev/c/amp-video-docking) instead.

File: extensions/amp-mowplayer/0.1/amp-mowplayer.js [hunk 3]
```diff
@@ -255,20 +286,237 @@ class AmpMowplayer extends AMP.BaseElement {
...
+    this.stickyDisableWorker_.call();
+  }
+
+  /**
+   * Set player height on resize
+   * @private
+   * @param {Object} data
+   */
+  onResize_(data) {
+    if (!window.mowStickyEnabled) {
```
Comment (alanorozco) [inline extensions/amp-mowplayer/0.1/amp-mowplayer.js:302]: Remove this flag.

File: extensions/amp-mowplayer/0.1/amp-mowplayer.js [hunk 4]
```diff
@@ -113,9 +129,19 @@ class AmpMowplayer extends AMP.BaseElement {
 
     installVideoManagerForDoc(this.element);
     Services.videoManagerForDoc(this.element).register(this);
+
+    const event = new CustomEvent('mowapiready', {
```
Comment (alanorozco) [inline extensions/amp-mowplayer/0.1/amp-mowplayer.js:133]: Why is this event dispatched at the document level?

File: extensions/amp-mowplayer/0.1/amp-mowplayer.js [hunk 5]
```diff
@@ -255,20 +286,237 @@ class AmpMowplayer extends AMP.BaseElement {
...
+    ) {
+      this.mofileFullApplied_ = true;
+      const repaint = () => {
+        const rect = this.element.getBoundingClientRect();
+        if (rect.width < window.screen.availWidth) {
+          const leftMargin = rect.left;
+          const rightMargin = window.screen.availWidth - rect.right;
+          setStyle(this.element, 'marginLeft', '-' + leftMargin + 'px');
+          setStyle(this.element, 'marginRight', '-' + rightMargin + 'px');
+        }
```
Comment (alanorozco) [inline extensions/amp-mowplayer/0.1/amp-mowplayer.js:328]: Please **don't** change sizing/positioning styles dynamically. **This causes content layout shift.**

**Comment by gopanear** [comment]
@googlebot I fixed it.

File: extensions/amp-mowplayer/0.1/amp-mowplayer.js (hunk 1 above)
Comment (gopanisandip) [reply extensions/amp-mowplayer/0.1/amp-mowplayer.js:519]: @alanorozco Let me know if is works

```
const worker = () => {
      const {intersectionRatio} = this.element.getIntersectionChangeEntry();
      const visible = intersectionRatio > 0.5 ? true : false;
      this.sendMessage_('visibility_observer_visibility', {visible});
};
```
We also need to pause autoplay when player is less than 50%  so we need to send Message to player.

File: extensions/amp-mowplayer/0.1/amp-mowplayer.js (hunk 2 above)
Comment (gopanisandip) [reply extensions/amp-mowplayer/0.1/amp-mowplayer.js:341]: @alanorozco  

Removed

amp-video-docking is something that our user need to add that componenet js and add doct attribute?   Right?   Or this is something that can be done within our componenet and user no need to worry about that?

File: extensions/amp-mowplayer/0.1/amp-mowplayer.js (hunk 3 above)
Comment (gopanisandip) [reply extensions/amp-mowplayer/0.1/amp-mowplayer.js:302]: @alanorozco Removed

File: extensions/amp-mowplayer/0.1/amp-mowplayer.js (hunk 4 above)
Comment (gopanisandip) [reply extensions/amp-mowplayer/0.1/amp-mowplayer.js:133]: @alanorozco  It was added in past but is not needed now i think. so i will remove it.

File: extensions/amp-mowplayer/0.1/amp-mowplayer.js (hunk 5 above)
Comment (gopanisandip) [reply extensions/amp-mowplayer/0.1/amp-mowplayer.js:328]: @alanorozco Removed

**Comment by gopanear** [comment]
@googlebot I fixed it.

**Comment by gopanisandip** [comment]
@googlebot I fixed it.

**Comment by gopanisandip** [comment]
@alanorozco    Closing this request due to CLA  issue and cretaed fresh PR [https://github.com/ampproject/amphtml/pull/29307](https://github.com/ampproject/amphtml/pull/29307)

**Pull request closed without merging**